src/Tests/Microsoft.NET.Sdk.Razor.SourceGenerators.Tests/RazorSourceGeneratorTests.cs

```diff
@@ -1164,4 +2664,68 @@ public override bool TryGetValue(string key, [NotNullWhen(true)] out string? val
                 => Options.TryGetValue(key, out value);
         }
     }
+
+    internal static class Extensions
+    {
+        public static GeneratorRunResult VerifyPageOutput(this GeneratorRunResult result, params string[] expectedOutput)
+        {
+            if (expectedOutput.Length == 1 && string.IsNullOrWhiteSpace(expectedOutput[0]))
+            {
+                GenerateExpectedOutput(result);
+            }
+            else
+            {
+                Assert.Equal(expectedOutput.Length, result.GeneratedSources.Length);
+                for (int i = 0; i < result.GeneratedSources.Length; i++)
+                {
+                    var text = result.GeneratedSources[i].SourceText.ToString();
+                    Assert.Equal(text.Trim('\r', '\n'), expectedOutput[i].Trim('\r', '\n'), ignoreWhiteSpaceDifferences: true);
+                }
+            }
+
+            return result;
+        }
+
+        private static void GenerateExpectedOutput(GeneratorRunResult result)
+        {
+            StringBuilder sb = new StringBuilder("Generated Output:").AppendLine().AppendLine();
+            for (int i = 0; i < result.GeneratedSources.Length; i++)
+            {
+                if (i > 0)
+                {
+                    sb.AppendLine(",");
+                }
+                sb.Append("@\"").Append(result.GeneratedSources[i].SourceText.ToString().Replace("\"", "\"\"")).Append('"');
+            }
+            Assert.True(false, sb.ToString());
+        }
+
+        public static GeneratorRunResult VerifyOutputsMatch(this GeneratorRunResult actual, GeneratorRunResult expected, params (int index, string replacement)[] diffs)
+        {
+            Assert.Equal(actual.GeneratedSources.Length, expected.GeneratedSources.Length);
+            for (int i = 0; i < actual.GeneratedSources.Length; i++)
+            {
+                var diff = diffs.FirstOrDefault(p => p.index == i).replacement;
+                if (diff is null)
+                {
+                    try
+                    {
+                        Assert.Equal(expected.GeneratedSources[i].SourceText.ToString(), actual.GeneratedSources[i].SourceText.ToString());
+                    }
+                    catch (EqualException e)
+                    {
+                        Assert.False(true, $"No diff supplied. But index {i} was:\r\n\r\n{e.Actual.Replace("\"", "\"\"")}");
+                    }
```

Comment:
What is the `try/catch` for? Could we use `Assert.True(expectedText == actualText, $"... index {i} was ... {actual}");` instead?

Reply:
Ah yeah, that was just from where I was iterating to get a decent mechanism for comparison. That path is only used if you don't supply a diff at all, so we can just assert if they don't match.